File: Strivers-A2Z-DSA-Sheet/04.Linked List/3.Medium Problems of LL/python/01.Find_mid_of_LL.py
"""
QUESTION:
Given the head of a singly linked list, return the middle node.
If there are two middle nodes, return the second middle node.

Example 1:  Input: [1,2,3,4,5]    Output: node(3)
Example 2:  Input: [1,2,3,4,5,6]  Output: node(4)

APPROACH: Slow & Fast Pointer (Floyd's tortoise)
- slow moves 1 step, fast moves 2 steps.
- When fast reaches the end, slow is at the middle.
- For even-length lists, slow lands on the 2nd middle.

TIME COMPLEXITY:  O(N)
SPACE COMPLEXITY: O(1)
"""


class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None


def middle_node(head):
    slow = head
    fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    return slow  # slow is at the middle


# Helpers
def build_ll(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    curr = head
    for i in range(1, len(arr)):
        curr.next = ListNode(arr[i]); curr = curr.next
    return head

def print_ll(head):
    result = []
    curr = head
    while curr:
        result.append(str(curr.val)); curr = curr.next
    print(" -> ".join(result))


# Driver
if __name__ == "__main__":
    print_ll(middle_node(build_ll([1,2,3,4,5])))    # 3 -> 4 -> 5
    print_ll(middle_node(build_ll([1,2,3,4,5,6])))  # 4 -> 5 -> 6

File: Strivers-A2Z-DSA-Sheet/04.Linked List/3.Medium Problems of LL/javascript/01.Find_mid_of_LL.js
/*
QUESTION:
Given the head of a singly linked list, return the middle node.
If there are two middle nodes, return the second middle node.

Example 1:
  Input:  [1,2,3,4,5]    Output: [3,4,5]   (node 3 is middle)
Example 2:
  Input:  [1,2,3,4,5,6]  Output: [4,5,6]   (two middles → return 2nd)

APPROACH: Slow & Fast Pointer (Floyd's tortoise)
- slow moves 1 step, fast moves 2 steps.
- When fast reaches the end, slow is exactly at the middle.
- For even-length lists fast stops at last node → slow lands on 2nd middle.

TIME COMPLEXITY:  O(N)
SPACE COMPLEXITY: O(1)
*/

class ListNode {
    constructor(val) { this.val = val; this.next = null; }
}

function middleNode(head) {
    let slow = head;
    let fast = head;

    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
    }

    return slow; // slow is at the middle
}

// Helpers
function buildLL(arr) {
    if (!arr.length) return null;
    const head = new ListNode(arr[0]);
    let curr = head;
    for (let i = 1; i < arr.length; i++) { curr.next = new ListNode(arr[i]); curr = curr.next; }
    return head;
}
function printLL(head) {
    const r = []; let c = head;
    while (c) { r.push(c.val); c = c.next; }
    console.log(r.join(' -> '));
}

// Driver
printLL(middleNode(buildLL([1,2,3,4,5])));   // 3 -> 4 -> 5
printLL(middleNode(buildLL([1,2,3,4,5,6]))); // 4 -> 5 -> 6

